In [1]:
using ExcelReaders
using PyPlot

In [3]:
f = openxl("sase-2016-hitorico-10032017.xlsx")

ExcelFile <sase-2016-hitorico-10032017.xlsx>

In [7]:
g = openxl("aeropuertos-septiembre-2017-23102017.xlsx")

ExcelFile <aeropuertos-septiembre-2017-23102017.xlsx>

In [42]:
DataNac = readxl(f, "REG NAC!A7:AO651");

In [126]:
DataInt = readxl(f, "REG INT!A7:AO868");

In [76]:
DataAeropuertos = readxl(g, "TD AEROPUERTOS!A7:M67");

### Pregunta 2

#### a) Saber el número de aeropuertos. 

Primero busquemos si el número de aeropuertos de donde salen aviones es el mismo que aquellos a donde llegan aviones.

#### Para vuelos nacionales:



DataNac[:, 1] se refiere a los aeropuertos de origen;

DataNac[:, 2] se refiere a los aeropuertos destino.

In [316]:
origen_destino = 1
count_aero = 0
for i = 2:length(DataNac[:, origen_destino])
    if sort(DataNac[:, origen_destino])[i-1] != sort(DataNac[:, origen_destino])[i]
        count_aero += 1
    end
end
count_aero

58

In [317]:
origen_destino = 2
count_aero = 0
for i = 2:length(DataNac[:, origen_destino])
    if sort(DataNac[:, origen_destino])[i-1] != sort(DataNac[:, origen_destino])[i]
        count_aero += 1
    end
end
count_aero

58

¡Check! Hay 58 aeropuertos en el país de donde salen y llegan vuelos nacionales a partir de la base de datos de origenes y destinos.

Calculemos cuantos aeropuertos se encuentran en ```DataAeropuertos```.

In [77]:
length(DataAeropuertos[:, end])

61

Hay una diferencia de dos aeropuertos...

Sin embargo, realizando un pequeño análisis acerca de los aeropuertos que tuvieron actividad en 2017 (año de las tablas de origen-destino), ¡encontramos que hay dos aeropuertos que no tuvieron actividad en 2017!

In [74]:
DataAeropuertos[60,:]

13-element DataArrays.DataArray{Any,1}:
  "TERAN"
 0.0     
 0.0     
 0.0     
 0.0     
 0.0     
 0.0     
 0.0     
 0.0     
 0.0     
 0.0     
 0.0     
 0.0     

In [78]:
DataAeropuertos[61, :]

13-element DataArrays.DataArray{Any,1}:
    "SAN CRISTOBAL DE LAS CASAS"
  29.0                          
  50.0                          
  83.0                          
 112.0                          
   0.0                          
   0.0                          
   0.0                          
   0.0                          
   0.0                          
   0.0                          
   0.0                          
   0.0                          

In [315]:
DataAeropuertos[52, :]

13-element DataArrays.DataArray{Any,1}:
     "IXTEPEC"
    0.0       
    0.0       
    0.0       
    0.0       
    0.0       
    0.0       
    0.0       
    0.0       
    0.0       
    0.0       
    0.0       
 2268.53      

Observamos como no hay ningún dato para el aeropuerto de Terán, San Cristobal de las Casas e Ixtepec en 2016...


### ¿Son consistentes las diferentes tablas entre esos 58 aeropuertos?

#### Encontramos que los siguientes aeropuertos solo están en la tabla de origen-destino

+ Jalapa;

+ Lázaro Cárdenas;

+ Piedras Negras;

+ Punta Peñasco;

+ Saltillo.

####  Y lo siguientes sólo se encuentran en la tabla de aeropuertos

+ Guaymas;

+ Nogales;

+ Tamuín;

+ Tehuacán.

#### Ya aquí encontramos un problema... ¿qué hacer con estos?

+ Según Wikipedia/Anexo: Estadísticas de aeropuertos de México (hm hm), tanto Guaymas, Nogales, Tamuín y Tehuacán no tienen vuelos comerciales, lo cual explicaría porque no están en la tabla de origen-destino.

+ Los cinco aeropuertos que no se encuentran en la tabla de aeropuertos son considerados como "aeropuertos pequeños" según Wikipedia.

+ Al leer las referencias en Wikipedia, encontramos que se citan a los mismo datos que se están analizando aquí. Esto vuelve al argumento un poco circular, pero para fines de rápidez y de utilidad, se mantendrá lo dicho.

#### Si para los "aeropuertos pequeños" hay estadísticas, entonces vamos a incluirlos en la lista de aeropuertos.

#### Estudiemos ahora los vuelos internacionales

#### Veámos si pasa lo mismo para vuelos internacionales.

In [127]:
origen_destino = 1
count_aero_int = 1
for i = 2:length(DataInt[:, 1])
    if sort(DataInt[:, origen_destino])[i-1] != sort(DataInt[:, origen_destino])[i]
        count_aero_int += 1
    end
end
count_aero_int

166

In [128]:
origen_destino = 2
count_aero_int = 1
for i = 2:length(DataInt[:, 1])
    if sort(DataInt[:, origen_destino])[i-1] != sort(DataInt[:, origen_destino])[i]
        count_aero_int += 1
    end
end
count_aero_int

176

Para ```origen_destino=1``` encontramos 166 aeropuertos, mientras que para ```origen_destino=2``` encontramos 176 aeropuertos... Esto significa que hay un mayor número de aeropuertos destino que de origen.

En 2017 solo hay 59 aeropuertos mexicanos activos. ¿Todos estos participan en vuelos internacionales?

#### nombres de los aeropuertos

In [355]:
Nombres = Array{String}(54)
Nombres[1:51] = DataAeropuertos[1:51,1];
Nombres[52:53] = DataAeropuertos[53:54, 1];
Nombres[54] = DataAeropuertos[56, 1];

+ Vamos a agregar los cinco "aeropuertos pequeños"

In [356]:
push!(Nombres, "JALAPA", "LAZARO CARDENAS", "PIEDRAS NEGRAS", "PUNTA PEÑASCO", "SALTILLO")

59-element Array{String,1}:
 "CIUDAD DE MÉXICO/MEXICO CITY"
 "CANCUN"                      
 "GUADALAJARA"                 
 "MONTERREY"                   
 "TIJUANA"                     
 "SAN JOSE DEL CABO"           
 "PUERTO VALLARTA"             
 "MERIDA"                      
 "BAJIO"                       
 "CULIACAN"                    
 "HERMOSILLO"                  
 "CHIHUAHUA"                   
 "VERACRUZ"                    
 ⋮                             
 "MATAMOROS"                   
 "NUEVO LAREDO"                
 "CD. VICTORIA"                
 "LORETO"                      
 "POZA RICA"                   
 "CUERNAVACA"                  
 "PALENQUE"                    
 "JALAPA"                      
 "LAZARO CARDENAS"             
 "PIEDRAS NEGRAS"              
 "PUNTA PEÑASCO"               
 "SALTILLO"                    

+ Notamos que mientras que en DataNac y DataInt, el Aeropuerto Internacional de la Ciudad de México es llamado por "MEXICO", en DataAeropuertos es llamado por "CIUDAD DE MÉXICO/MEXICO CITY";

+ También notamos que en DataNac y DataInt, se encuentra "SAN JOSÉ DEL CABO", en DataAeropuertos se encuentra "SAN JOSE DEL CABO";

+ "TUXTLA GUTIERREZ (ANGEL ALBINO CORZO)" se encuentra en las tablas de origen destino como "TUXTLA GUTIERREZ";

+ Las Ciudades de Juárez/Obregón/Victoria/Del Carmen se encuentran como "CD. ---" y como "CIUDAD ---";

+ "QUERÉTARO" se diferencia de "QUERETARO";

+ "ZIHUATANEJO" se encuentra como "IXTAPA ZIHUATANEJO";

+ Finalmente, "BAJIO" también se encuentra como "DEL BAJIO".

In [82]:
Nombres[1]

"CIUDAD DE MÉXICO/MEXICO CITY"

In [83]:
DataNac[3, 2]

"MEXICO"

Vamos a cambiar todos los errores (pff....).

In [357]:
Nombres[1] = "MEXICO";

In [358]:
Nombres[6] = "SAN JOSÉ DEL CABO"

"SAN JOSÉ DEL CABO"

In [359]:
Nombres[14] = "TUXTLA GUTIERREZ"

"TUXTLA GUTIERREZ"

In [360]:
Nombres[15] = "CIUDAD JUAREZ"


"CIUDAD JUAREZ"

In [361]:
Nombres[33] = "CIUDAD DEL CARMEN"

"CIUDAD DEL CARMEN"

In [362]:
Nombres[36] = "CIUDAD OBREGON"

"CIUDAD OBREGON"

In [363]:
Nombres[50] = "CIUDAD VICTORIA"

"CIUDAD VICTORIA"

In [364]:
Nombres[18] = "QUERETARO"

"QUERETARO"

In [365]:
Nombres[35] = "IXTAPA ZIHUATANEJO"

"IXTAPA ZIHUATANEJO"

In [366]:
Nombres[9] = "DEL BAJIO"

"DEL BAJIO"

Ahora sí... veamos qué aeropuertos nacionales están en los registros internacionales.

#### Como aeropuerto de salida

In [367]:
Aeropuertos_Int_Salida = Array{String}(0)
origen_destino = 1

for i = 2:length(DataInt[:, origen_destino])
    if sort(DataInt[:, origen_destino])[i-1] != sort(DataInt[:, origen_destino])[i]
        push!(Aeropuertos_Int_Salida, sort(DataInt[:, origen_destino])[i-1])
    end
end

if Aeropuertos_Int_Salida[end] != sort(DataInt[:, origen_destino])[end]
    push!(Aeropuertos_Int_Salida, sort(DataInt[:, origen_destino])[end])
end

166-element Array{String,1}:
 "ABBOTSFORD"            
 "ACAPULCO"              
 "AGUASCALIENTES"        
 "AMSTERDAM"             
 "ATLANTA"               
 "AUSTIN"                
 "BAGOTVILLE"            
 "BALTIMORE"             
 "BELICE"                
 "BIRMINGHAM, INGLATERRA"
 "BOGOTA"                
 "BOSTON"                
 "BROWNSVILLE"           
 ⋮                       
 "VANCOUVER"             
 "VARADERO"              
 "VERACRUZ"              
 "VERONA"                
 "VICTORIA, COLUMBIA"    
 "VILLAHERMOSA"          
 "WASHINGTON"            
 "WINDSOR"               
 "WINDSOR LOCKS"         
 "WINNIPEG"              
 "ZACATECAS"             
 "ZARAGOZA, ES"          

In [371]:
Aeropuertos_NacEInt_Salidas = Array{String}(0)

for aeropuerto in Nombres
    if any(x->x==aeropuerto, Aeropuertos_Int_Salida)
        push!(Aeropuertos_NacEInt_Salidas, aeropuerto)
    end
end

Aeropuertos_NacEInt_Salidas

37-element Array{String,1}:
 "MEXICO"            
 "CANCUN"            
 "GUADALAJARA"       
 "MONTERREY"         
 "TIJUANA"           
 "SAN JOSÉ DEL CABO" 
 "PUERTO VALLARTA"   
 "MERIDA"            
 "DEL BAJIO"         
 "CULIACAN"          
 "HERMOSILLO"        
 "CHIHUAHUA"         
 "VERACRUZ"          
 ⋮                   
 "MORELIA"           
 "ACAPULCO"          
 "DURANGO"           
 "CIUDAD DEL CARMEN" 
 "ZACATECAS"         
 "IXTAPA ZIHUATANEJO"
 "COZUMEL"           
 "TAPACHULA"         
 "LOS MOCHIS"        
 "URUAPAN"           
 "MANZANILLO"        
 "LORETO"            

#### Como aeropuerto destino

In [369]:
Aeropuertos_Int_Destino = Array{String}(0)
origen_destino = 2

for i = 2:length(DataInt[:, origen_destino])
    if sort(DataInt[:, origen_destino])[i-1] != sort(DataInt[:, origen_destino])[i]
        push!(Aeropuertos_Int_Destino, sort(DataInt[:, origen_destino])[i-1])
    end
end

if Aeropuertos_Int_Destino[end] != sort(DataInt[:, origen_destino])[end]
    push!(Aeropuertos_Int_Destino, sort(DataInt[:, origen_destino])[end])
end

176-element Array{String,1}:
 "ABBOTSFORD"            
 "ACAPULCO"              
 "AGUASCALIENTES"        
 "AMSTERDAM"             
 "ATLANTA"               
 "AUSTIN"                
 "Antofagasta"           
 "BALTIMORE"             
 "BELICE"                
 "BIRMINGHAM, INGLATERRA"
 "BOGOTA"                
 "BOSTON"                
 "BROWNSVILLE"           
 ⋮                       
 "VANCOUVER"             
 "VARADERO"              
 "VERACRUZ"              
 "VERONA"                
 "VICTORIA, COLUMBIA"    
 "VILLAHERMOSA"          
 "WASHINGTON"            
 "WINDSOR"               
 "WINDSOR LOCKS"         
 "WINNIPEG"              
 "XIAMEN"                
 "ZACATECAS"             

In [370]:
Aeropuertos_NacEInt_Destino = Array{String}(0)

for aeropuerto in Nombres
    if any(x->x==aeropuerto, Aeropuertos_Int_Destino)
        push!(Aeropuertos_NacEInt_Destino, aeropuerto)
    end
end

Aeropuertos_NacEInt_Destino

38-element Array{String,1}:
 "MEXICO"            
 "CANCUN"            
 "GUADALAJARA"       
 "MONTERREY"         
 "TIJUANA"           
 "SAN JOSÉ DEL CABO" 
 "PUERTO VALLARTA"   
 "MERIDA"            
 "DEL BAJIO"         
 "CULIACAN"          
 "HERMOSILLO"        
 "CHIHUAHUA"         
 "VERACRUZ"          
 ⋮                   
 "ACAPULCO"          
 "DURANGO"           
 "CIUDAD DEL CARMEN" 
 "ZACATECAS"         
 "IXTAPA ZIHUATANEJO"
 "CIUDAD OBREGON"    
 "COZUMEL"           
 "TAPACHULA"         
 "LOS MOCHIS"        
 "URUAPAN"           
 "MANZANILLO"        
 "LORETO"            

+ Sin embargo, encontramos que el aeropuerto de Ciudad Juárez salen vuelos internacionales, pero no regresan.

+ De los aeropuertos de Mexicali y Ciudad Obregón, hay vuelos internacionales que llegan a esos aeropuertos, pero no regresan a sus destino.

#### En conclusión

+ Hay 37 aeropuertos nacionales de donde salen vuelos internacionales;
+ Hay 38 aeropuertos nacionales a donde llegan vuelos internacionales;
+ Hay 39 aeropuertos nacionales de donde salen o llegan vuelos internacionales;
+ Contando TODOS los aeropuertos con estadísticas de pasajeros, de los 63 aeropuertos que hay (59 con estadísticas de vuelos + 4 sin estadísticas de vuelo), 24 aeropuertos no registraron vuelos internacionales;
+ Hay 129 aeropuertos internacionales que tienen vuelos hacia México;
+ Hay 138 aeropuertos internacionales que reciben vuelos desde México.

Llama la atención la no simetría entre los aeropuertos internacionales que reciben vuelos de México pero desde donde no se vuela a México.

### 2-b

A partir de las tablas de origen y destino, lo que se mide es la cantidad de vuelos de un lugar a otro, el número de pasajeros y los kilos de carga. Estas tres medidas son mensuales.

En un primer momento se puede saber la cantidad promedio de pasajeros por vuelo. Esto nos puede dar una idea del tipo de avión que se ocupa para los vuelos entre una ciudad y otra. Esta información puede ser sensible para las aerolíneas que buscan expanderse hacia otras ciudades o que quieran informarse acerca de qué tipo de aeronave es conveniente para el mercado en México.

### 2-c

### Flujos de entrada y de salida

Aquí tendremos que hacer una pequeña especificación acerca del flujo. ¿Estamos hablando de flujo de aviones, o de personas?¿Flujo mensual, o anual?

Si queremos saber la capacidad de vuelos de un aeropuerto, para saber si por ejemplo el número de pistas está llegando a su límite, entonces nos decantaríamos por flujo de aviones. Si por el otro lado, queremos si la capacidad de personas del aeropuerto está llegando a su límite, entonces vale más la pena usar un flujo de pasajeros.

En este caso me decantaré por un flujo anual. Esto debido a que encontramos una cantidad considerable de meses en los cuales hay cero vuelos entre dos ciudades.

También utilizaré un flujo de personas, ya que me parece más relevante. Si hay un gran flujo de personas, entonces se pueden cambiar los aviones por algunos con más capacidad y modificar la infraestructura aeropuertaria esperando buenos resultados. Si hay una gran frecuencia de vuelos, esto no garantiza que el flujo de personas pueda aumentar, o que valga la pena cambiar los aviones por unos de mayor capacidad.

(además creo que ya me estoy extendiendo mucho)

#### Sumar flujos de salida de pasajeros nacionales e internacionales

In [388]:
origen_destino = 1

Flujo_Salida = Dict{String, Float64}() 

for aeropuerto in Nombres
    for i = 1:length(DataNac[:, origen_destino])
        if DataNac[i, origen_destino] == aeropuerto
            if haskey(Flujo_Salida, aeropuerto)
                Flujo_Salida[aeropuerto] += DataNac[i, 28]
            else 
                Flujo_Salida[aeropuerto] = DataNac[i, 28]
            end
        end
    end
end

In [389]:
Flujo_Salida["MEXICO"]

1.3872101e7

In [390]:
origen_destino = 1

for aeropuerto in Aeropuertos_NacEInt_Salidas
    for i = 1:length(DataInt[:, origen_destino])
        if DataInt[i, origen_destino] == aeropuerto
            Flujo_Salida[aeropuerto] += DataInt[i, 28]
        end
    end
end

In [391]:
Flujo_Salida["MEXICO"]

2.0769176e7

#### Sumar flujos de entrada de pasajeros nacionales e internacionales

In [392]:
origen_destino = 2

Flujo_Entrada = Dict{String, Float64}() 

for aeropuerto in Nombres
    for i = 1:length(DataNac[:, origen_destino])
        if DataNac[i, origen_destino] == aeropuerto
            if haskey(Flujo_Entrada, aeropuerto)
                Flujo_Entrada[aeropuerto] += DataNac[i, 28]
            else 
                Flujo_Entrada[aeropuerto] = DataNac[i, 28]
            end
        end
    end
end

In [393]:
Flujo_Entrada["MEXICO"]

1.3864686e7

In [394]:
origen_destino = 2

for aeropuerto in Aeropuertos_NacEInt_Salidas
    for i = 1:length(DataInt[:, origen_destino])
        if DataInt[i, origen_destino] == aeropuerto
            Flujo_Entrada[aeropuerto] += DataInt[i, 28]
        end
    end
end

In [395]:
Flujo_Entrada["MEXICO"]

2.0874275e7

Ahora podemos saber los flujos totales (entrada + salida) y las diferencias de flujo (entrada - salida)

In [396]:
Flujo_Total = Dict{String, Float64}()
Flujo_Dif = Dict{String, Float64}()

for aeropuerto in Nombres
    Flujo_Total[aeropuerto] = Flujo_Entrada[aeropuerto] + Flujo_Salida[aeropuerto]
    Flujo_Dif[aeropuerto] = Flujo_Entrada[aeropuerto] - Flujo_Salida[aeropuerto]
end

+ Flujo_Total nos da el número neto de pasajeros que pisaron el aeropuerto.
+ Flujo_Dif, si bien nos da una idea si salieron más o entraron más pasajeros, no nos da la perspectiva completa (¿qué son 50k pasajeros si hubo 50 millones?)

Hagamos una representación proporcional con respecto al flujo total de pasajeros

In [401]:
Flujo_Prop = Dict{String, Float64}()

for aeropuerto in Nombres
    Flujo_Prop[aeropuerto] = 100*Flujo_Dif[aeropuerto]/Flujo_Total[aeropuerto]
    end                                                    

En porcentaje con respecto al numéro total de pasajeros que cruzaron el aeropuerto. Si el resultado es negativo, es porque salieron más pasajeros al aeropuerto de los que entraron.

In [402]:
Flujo_Prop 

Dict{String,Float64} with 59 entries:
  "CANCUN"             => -0.313831
  "ACAPULCO"           => -1.55873
  "PUERTO ESCONDIDO"   => -2.43428
  "GUADALAJARA"        => -0.261923
  "MEXICALI"           => 2.63872
  "TIJUANA"            => 1.47852
  "AGUASCALIENTES"     => -0.525017
  "QUERETARO"          => -0.463273
  "COLIMA"             => 0.29219
  "SALTILLO"           => 2.99704
  "PUNTA PEÑASCO"      => -5.75831
  "MATAMOROS"          => 5.14901
  "TAPACHULA"          => -5.54774
  "PIEDRAS NEGRAS"     => 4.42986
  "LAZARO CARDENAS"    => -0.543751
  "POZA RICA"          => -1.28732
  "NUEVO LAREDO"       => 10.2785
  "TUXTLA GUTIERREZ"   => 0.505303
  "IXTAPA ZIHUATANEJO" => -0.293927
  "LOS MOCHIS"         => -1.23194
  "ZACATECAS"          => -0.938589
  "TORREON"            => -0.385447
  "PALENQUE"           => -6.65744
  "CULIACAN"           => -0.681561
  "SAN JOSÉ DEL CABO"  => 0.0642885
  ⋮                    => ⋮

#### Hagámoslo ordenadamente

In [430]:
Orden_P = sort(collect(Flujo_Total), by=x->x[2], rev=true)

59-element Array{Pair{String,Float64},1}:
 "MEXICO"=>4.16435e7           
 "CANCUN"=>1.97593e7           
 "GUADALAJARA"=>1.12251e7      
 "MONTERREY"=>8.93421e6        
 "TIJUANA"=>6.33059e6          
 "SAN JOSÉ DEL CABO"=>4.05516e6
 "PUERTO VALLARTA"=>3.75098e6  
 "MERIDA"=>1.94027e6           
 "DEL BAJIO"=>1.69286e6        
 "CULIACAN"=>1.67395e6         
 "HERMOSILLO"=>1.53121e6       
 "CHIHUAHUA"=>1.32915e6        
 "VERACRUZ"=>1.31353e6         
 ⋮                             
 "MATAMOROS"=>75801.0          
 "NUEVO LAREDO"=>74729.0       
 "LORETO"=>74191.0             
 "SALTILLO"=>65598.0           
 "CIUDAD VICTORIA"=>50922.0    
 "POZA RICA"=>32393.0          
 "LAZARO CARDENAS"=>23908.0    
 "CUERNAVACA"=>22759.0         
 "PIEDRAS NEGRAS"=>20565.0     
 "PALENQUE"=>13008.0           
 "JALAPA"=>3708.0              
 "PUNTA PEÑASCO"=>2466.0       

Claramente se ve que el aeropuerto que recibe más pasajeros es el AICM, mientras que el que menos recibe es Punta Peñasco.

#### Ahora en cuanto a cuales varían más su porcentaje de salidas contra entradas

In [416]:
sort(collect(Flujo_Prop), by=x->x[2], rev=true)

59-element Array{Pair{String,Float64},1}:
 "NUEVO LAREDO"=>10.2785      
 "JALAPA"=>5.87918            
 "MATAMOROS"=>5.14901         
 "PIEDRAS NEGRAS"=>4.42986    
 "REYNOSA"=>4.03098           
 "SALTILLO"=>2.99704          
 "MEXICALI"=>2.63872          
 "CIUDAD JUAREZ"=>2.13764     
 "TIJUANA"=>1.47852           
 "MERIDA"=>1.29637            
 "CHETUMAL"=>0.815599         
 "HERMOSILLO"=>0.793946       
 "CIUDAD VICTORIA"=>0.655905  
 ⋮                            
 "LOS MOCHIS"=>-1.23194       
 "POZA RICA"=>-1.28732        
 "VILLAHERMOSA"=>-1.48078     
 "ACAPULCO"=>-1.55873         
 "MINATITLAN"=>-1.9077        
 "PUERTO ESCONDIDO"=>-2.43428 
 "URUAPAN"=>-2.48349          
 "CIUDAD DEL CARMEN"=>-2.89314
 "CUERNAVACA"=>-3.12404       
 "TAPACHULA"=>-5.54774        
 "PUNTA PEÑASCO"=>-5.75831    
 "PALENQUE"=>-6.65744         

In [421]:
sort(collect(Flujo_Prop), by=x->x[2], rev=true)[25:30]

6-element Array{Pair{String,Float64},1}:
 "DEL BAJIO"=>0.0926834        
 "SAN JOSÉ DEL CABO"=>0.0642885
 "PUERTO VALLARTA"=>0.0424955  
 "HUATULCO"=>-0.0433149        
 "DURANGO"=>-0.0539084         
 "MANZANILLO"=>-0.0865279      

+ Nuevo Laredo es el aeropuerto donde hay una variación más fuerte: Hay una variación de poco más del 10% de pasajeros que entran con respecto a los que salen.
+ Palenque, por el otro lado, tiene una variación del 6.66% de pasajeros que salen conr especto a los que entren. 
+ Puerto Vallarta y Huatuclo son los que tienen una menor variación. Prácticamente entra el mismo número de pasajeros de los que salen.

#### El siguiente chorizo es para los siguiente rubros. Pueden irse hasta los resultados (puede que haya sido más fácil hacer una función, pero en fin).

In [424]:
#### Vuelos

## Salidas
origen_destino = 1

Flujo_Salida_Vuelos = Dict{String, Float64}() 

for aeropuerto in Nombres
    for i = 1:length(DataNac[:, origen_destino])
        if DataNac[i, origen_destino] == aeropuerto
            if haskey(Flujo_Salida_Vuelos, aeropuerto)
                Flujo_Salida_Vuelos[aeropuerto] += DataNac[i, 15]
            else 
                Flujo_Salida_Vuelos[aeropuerto] = DataNac[i, 15]
            end
        end
    end
end

for aeropuerto in Aeropuertos_NacEInt_Salidas
    for i = 1:length(DataInt[:, origen_destino])
        if DataInt[i, origen_destino] == aeropuerto
            Flujo_Salida_Vuelos[aeropuerto] += DataInt[i, 15]
        end
    end
end

## Entradas
origen_destino = 2

Flujo_Entrada_Vuelos = Dict{String, Float64}() 

for aeropuerto in Nombres
    for i = 1:length(DataNac[:, origen_destino])
        if DataNac[i, origen_destino] == aeropuerto
            if haskey(Flujo_Entrada_Vuelos, aeropuerto)
                Flujo_Entrada_Vuelos[aeropuerto] += DataNac[i, 15]
            else 
                Flujo_Entrada_Vuelos[aeropuerto] = DataNac[i, 15]
            end
        end
    end
end

for aeropuerto in Aeropuertos_NacEInt_Salidas
    for i = 1:length(DataInt[:, origen_destino])
        if DataInt[i, origen_destino] == aeropuerto
            Flujo_Entrada_Vuelos[aeropuerto] += DataInt[i, 15]
        end
    end
end


In [425]:
#### Carga

## Salidas
origen_destino = 1

Flujo_Salida_Carga = Dict{String, Float64}() 

for aeropuerto in Nombres
    for i = 1:length(DataNac[:, origen_destino])
        if DataNac[i, origen_destino] == aeropuerto
            if haskey(Flujo_Salida_Carga, aeropuerto)
                Flujo_Salida_Carga[aeropuerto] += DataNac[i, end]
            else 
                Flujo_Salida_Carga[aeropuerto] = DataNac[i, end]
            end
        end
    end
end

for aeropuerto in Aeropuertos_NacEInt_Salidas
    for i = 1:length(DataInt[:, origen_destino])
        if DataInt[i, origen_destino] == aeropuerto
            Flujo_Salida_Carga[aeropuerto] += DataInt[i, end]
        end
    end
end

## Entradas
origen_destino = 2

Flujo_Entrada_Carga = Dict{String, Float64}() 

for aeropuerto in Nombres
    for i = 1:length(DataNac[:, origen_destino])
        if DataNac[i, origen_destino] == aeropuerto
            if haskey(Flujo_Entrada_Carga, aeropuerto)
                Flujo_Entrada_Carga[aeropuerto] += DataNac[i, end]
            else 
                Flujo_Entrada_Carga[aeropuerto] = DataNac[i, end]
            end
        end
    end
end

for aeropuerto in Aeropuertos_NacEInt_Salidas
    for i = 1:length(DataInt[:, origen_destino])
        if DataInt[i, origen_destino] == aeropuerto
            Flujo_Entrada_Carga[aeropuerto] += DataInt[i, end]
        end
    end
end

In [426]:
Flujo_Total_Vuelos = Dict{String, Float64}()
Flujo_Prop_Vuelos = Dict{String, Float64}()

for aeropuerto in Nombres
    Flujo_Total_Vuelos[aeropuerto] = Flujo_Entrada_Vuelos[aeropuerto]+
                                                            Flujo_Salida_Vuelos[aeropuerto]
    Flujo_Prop_Vuelos[aeropuerto] = 100*(Flujo_Entrada_Vuelos[aeropuerto]-
                                                            Flujo_Salida_Vuelos[aeropuerto])/
                                                                Flujo_Total_Vuelos[aeropuerto]
end

Flujo_Total_Carga = Dict{String, Float64}()
Flujo_Prop_Carga = Dict{String, Float64}()

for aeropuerto in Nombres
    Flujo_Total_Carga[aeropuerto] = Flujo_Entrada_Carga[aeropuerto]+
                                                                Flujo_Salida_Carga[aeropuerto]
    Flujo_Prop_Carga[aeropuerto] = 100*(Flujo_Entrada_Carga[aeropuerto]-
                                                            Flujo_Salida_Carga[aeropuerto])/
                                                                Flujo_Total_Carga[aeropuerto]
end

#### Resultados

In [431]:
Orden_V = sort(collect(Flujo_Total_Vuelos), by=x->x[2], rev=true)

59-element Array{Pair{String,Float64},1}:
 "MEXICO"=>407305.0          
 "CANCUN"=>141393.0          
 "GUADALAJARA"=>106390.0     
 "MONTERREY"=>102520.0       
 "TIJUANA"=>50416.0          
 "SAN JOSÉ DEL CABO"=>33502.0
 "PUERTO VALLARTA"=>32746.0  
 "DEL BAJIO"=>23232.0        
 "HERMOSILLO"=>21424.0       
 "MERIDA"=>20426.0           
 "CHIHUAHUA"=>18674.0        
 "CULIACAN"=>16862.0         
 "VERACRUZ"=>16577.0         
 ⋮                           
 "MATAMOROS"=>1916.0         
 "CHETUMAL"=>1633.0          
 "NUEVO LAREDO"=>1315.0      
 "SALTILLO"=>1207.0          
 "LAZARO CARDENAS"=>1194.0   
 "LORETO"=>1092.0            
 "URUAPAN"=>972.0            
 "CUERNAVACA"=>737.0         
 "PIEDRAS NEGRAS"=>724.0     
 "JALAPA"=>365.0             
 "PALENQUE"=>214.0           
 "PUNTA PEÑASCO"=>141.0      

En cuanto a número de vuelos, el orden de los aeropuertos que tienen mayor y menor flujo son consistentes con el orden con respecto al flujo de pasajeros.

In [432]:
Orden_C = sort(collect(Flujo_Total_Carga), by=x->x[2], rev=true)

59-element Array{Pair{String,Float64},1}:
 "MEXICO"=>4.65475e8         
 "GUADALAJARA"=>1.66757e8    
 "MONTERREY"=>3.53908e7      
 "TOLUCA"=>2.05992e7         
 "TIJUANA"=>2.04368e7        
 "SAN LUIS POTOSI"=>1.93752e7
 "CANCUN"=>1.91994e7         
 "MERIDA"=>1.53179e7         
 "HERMOSILLO"=>8.57947e6     
 "CULIACAN"=>4.51275e6       
 "CHIHUAHUA"=>3.91057e6      
 "ACAPULCO"=>3.19109e6       
 "OAXACA"=>3.09553e6         
 ⋮                           
 "COZUMEL"=>91631.8          
 "LAZARO CARDENAS"=>82003.0  
 "MANZANILLO"=>78048.8       
 "SALTILLO"=>37414.5         
 "PUERTO ESCONDIDO"=>37265.0 
 "PIEDRAS NEGRAS"=>30052.0   
 "JALAPA"=>17348.0           
 "URUAPAN"=>2863.5           
 "LORETO"=>1797.0            
 "PUNTA PEÑASCO"=>0.0        
 "PALENQUE"=>0.0             
 "CUERNAVACA"=>0.0           

### Pregunta 3

+ Con respecto a Carga, encontramos que Cancún desciende un número de posiciones. Esto es congruente ya que Cancún se caracteriza por viajes de turismo y no por centro de distribución (al menos de cosas legales).
+ Con respecto a número de pasajeros, Querétaro desciende un gran número de posiciones. Esto significa que hay muchos vuelos pequeños que van a Querétaro.

#### 3-b

Me llama la atención la palabra "operación". Operación se refiere a, ¿un pasajero abordando?, ¿un kilo declarado?, ¿un avión tomando pista?

Si se toman esas definiciones como operaciones, entonces se pueden clasificar por los ordenes de magnitud. Por ejemplo, aeropuertos grandes son aquellos que tienen mas de un millón de pasajeros al año y aeropuertos pequeños aquellos que tienen menos de cincuenta mil.

Sin embargo estas clasificaciones no me encantan porque son a todas luces arbitrarias. Yo personalmente propondría una clasificación con respecto a la DENSIDAD. 

por ejemplo: densidad de un vuelo con respecto a sus pasajeros o a sus cargas, i.e. pasajeros/vuelo y kg/vuelo. Los datos ya están listos para hacer eso.

In [455]:
Densidad_Pasajeros = Dict{String, Float64}()
Densidad_Carga = Dict{String, Float64}()

for aeropuerto in Nombres
    Densidad_Pasajeros[aeropuerto] = Flujo_Total[aeropuerto]/Flujo_Total_Vuelos[aeropuerto]
    Densidad_Carga[aeropuerto] = Flujo_Total_Carga[aeropuerto]/Flujo_Total_Vuelos[aeropuerto]
end

In [456]:
Orden_DP = sort(collect(Densidad_Pasajeros), by=x->x[2], rev=true)

59-element Array{Pair{String,Float64},1}:
 "CANCUN"=>139.748           
 "CHETUMAL"=>125.688         
 "TIJUANA"=>125.567          
 "URUAPAN"=>121.668          
 "SAN JOSÉ DEL CABO"=>121.042
 "MEXICALI"=>118.303         
 "PUERTO VALLARTA"=>114.548  
 "COZUMEL"=>106.891          
 "GUADALAJARA"=>105.509      
 "MEXICO"=>102.241           
 "TAPACHULA"=>100.02         
 "CULIACAN"=>99.2736         
 "TUXTLA GUTIERREZ"=>99.2014 
 ⋮                           
 "LOS MOCHIS"=>49.4205       
 "TEPIC"=>42.1259            
 "COLIMA"=>40.2639           
 "MATAMOROS"=>39.5621        
 "SAN LUIS POTOSI"=>37.738   
 "CUERNAVACA"=>30.8806       
 "PIEDRAS NEGRAS"=>28.4047   
 "CIUDAD VICTORIA"=>21.7895  
 "LAZARO CARDENAS"=>20.0235  
 "PUNTA PEÑASCO"=>17.4894    
 "POZA RICA"=>13.5479        
 "JALAPA"=>10.1589           

In [458]:
Orden_DP = sort(collect(Densidad_Carga), by=x->x[2], rev=true)

59-element Array{Pair{String,Float64},1}:
 "TOLUCA"=>1806.0           
 "GUADALAJARA"=>1567.41     
 "SAN LUIS POTOSI"=>1538.2  
 "MEXICO"=>1142.82          
 "MERIDA"=>749.92           
 "MEXICALI"=>444.477        
 "TIJUANA"=>405.364         
 "HERMOSILLO"=>400.461      
 "MONTERREY"=>345.209       
 "OAXACA"=>320.017          
 "ACAPULCO"=>310.78         
 "CULIACAN"=>267.628        
 "TAPACHULA"=>260.945       
 ⋮                          
 "SALTILLO"=>30.9979        
 "MANZANILLO"=>22.2045      
 "ZACATECAS"=>22.1781       
 "COZUMEL"=>21.7653         
 "MORELIA"=>18.4424         
 "PUERTO ESCONDIDO"=>14.4214
 "QUERETARO"=>11.3442       
 "URUAPAN"=>2.94599         
 "LORETO"=>1.6456           
 "PUNTA PEÑASCO"=>0.0       
 "PALENQUE"=>0.0            
 "CUERNAVACA"=>0.0          

##  Estos resultados sí están interesantes. Vemos cómo para pasajeros, Cancún desplaza al AICM, puesto que los aviones vuelan, en promedio, con un mayor número de personas. Con respecto a la carga, tampoco el AICM se encuentra en la primera posición y es desplazado por Toluca. 

Esto nos permite clasificar los aeropuertos por sus densidades. Yo propondría cuatro categorías:
+ Turístico/Negocios. Muchos pasajeros, poca carga, como Cancún.
+ Centro de distribución. Pocos pasajeros, mucha carga, como Toluca.
+ Mixto. Muchos pasajeros y carga, como el AICM o Guadalajara.
+ Privado/local. Pocos pasajeros o poca carga, como Cuernavaca.

Con la densidad también podemos saber más fácilmente los periodos de operación, al igual que la diferencia entre operaciones nacionales e internacionales. Esto ya no me dio tiempo, pero dejo una lista de lo que hubiera hecho:

Hacer un análisis de tres series de tiempo para cada aeropuerto:
+ Número  mensual de vuelos nacionales e internacionales;
+ Densidad mensual de carga nacional e internacional;
+ Densidad mensual de pasajeros nacional e internacional.

Haciendo un skimming rápido de las tablas de Excel, una primera hipótesis sería que:

+ Los aeropuertos de turismo (Cancún, San José del Cabo, Puerto Vallarta) tienen un flujo muy pequeño de carga y muy grande de pasajeros. El flujo de pasajeros no es homogéneo y encuentra máximos en meses asociados a vacaciones, tales como verano, navidad y semana santa.

+ Los aeropuertos pequeños (Jalapa, Cuernavaca) tienen también poco flujo de carga y el flujo de pasajeros es bastante irregular, y no tiene una lógica clara.

+ Los aeropuertos de carga y mixtos (AICM, Guadalajara, Monterrey, Toluca) tienen una flujo constante a lo largo del año, tanto para pasajeros y de carga. Probablemente los mixtos tengan picos (aunque menos marcados que para aeropuertos de turismo) en temporadas altas de turismo debido al transborde de pasajeros. También no se me haría una locura que los aeropuertos de carga tengan picos semestrales, habiendo un descenso en la actividad durante los meses primavera-verano y un aumento de actividad durante otoño-invierno, en especial la internacional.